[NRBF] More bug fixes

adamsitnik opened the pull request:
Fixes issues discovered offline by @GrabYourPitchforks

adamsitnik committed:
be explicit, always throw InvalidOperationException for values not expected by switch statements and if/else blocks

adamsitnik committed:
be explicit, but also provide missing values

adamsitnik committed:
reject all arrays with length > Arraym.MaxLength (not just uint.MaxValue).

This is something I agreed on with Jeremy Kuhne a while ago, but missed it.

adamsitnik committed:
throw EOF on character count mismatch

adamsitnik committed:
remove decimal parsing optimization that was violating the NRBF spec

adamsitnik committed:
avoid int32 overflow

adamsitnik committed:
stop using Debug.Fail, remove all usages

adamsitnik committed:
StartsWithPayloadHeader should not throw on EOF

adamsitnik committed:
SerializationRecordId does not override ToString, we need to provide the value of the Id in explicit way

adamsitnik committed:
provide DebuggerDisplay for SerializationRecordId, fix it for ArrayInfo (Length property is longer present in this type)

adamsitnik commented:
@MihuBot fuzz NrbfDecoder

adamsitnik committed:
throw SerializationException when PrimitiveType.String is provided in a place where it's not valid

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/ArrayInfo.cs:
```diff
@@ -13,7 +13,7 @@ namespace System.Formats.Nrbf;
 /// <remarks>
 /// ArrayInfo structures are described in <see href="https://learn.microsoft.com/openspecs/windows_protocols/ms-nrbf/8fac763f-e46d-43a1-b360-80eb83d2c5fb">[MS-NRBF] 2.4.2.1</see>.
 /// </remarks>
-[DebuggerDisplay("Length={Length}, {ArrayType}, rank={Rank}")]
```
This type no longer provides `Length` property.

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/ArraySinglePrimitiveRecord.cs:
```diff
@@ -94,7 +86,7 @@ internal static IReadOnlyList<T> DecodePrimitiveTypes(BinaryReader reader, int c
         reader.BaseStream.ReadExactly(resultAsBytes);
 #else
-        byte[] bytes = ArrayPool<byte>.Shared.Rent(Math.Min(count * Unsafe.SizeOf<T>(), 256_000));
+        byte[] bytes = ArrayPool<byte>.Shared.Rent((int)Math.Min(requiredBytes, 256_000));
```
this fixes an int32 overflow: `large number of elements * sizeof(long)`

JeremyKuhne replied:
Why does `requiredBytes` not have a similar problem? Shouldn't that calculation be checked?

adamsitnik replied:
When it's calculated:

https://github.com/dotnet/runtime/blob/a3365e4c442a728bc82e59d5583cf83bc6d31d72/src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/ArraySinglePrimitiveRecord.cs#L64-L67

We know that count is less than `Array.MaxLength` and `T` is limited to a list of natively supported primitive types. So the max value we can get here is `Array.MaxLength * 8` which won't overflow `Int64`. Then `Math.Min(long,  256_000)` is going to return `256_000` which can be caster to `Int32` without an overflow.

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/ArraySinglePrimitiveRecord.cs:
```diff
@@ -159,31 +151,10 @@ internal static IReadOnlyList<T> DecodePrimitiveTypes(BinaryReader reader, int c
-
-            values.Add(value);
-        }
-#else
```
this perf optimization was wrong because it was not checking if given string input is a valid utf8 string

adamsitnik commented:
@MihuBot fuzz NrbfDecoder

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/BinaryArrayRecord.cs:
```diff
@@ -138,7 +138,7 @@ internal static ArrayRecord Decode(BinaryReader reader, RecordMap recordMap, Pay
             totalElementCount *= lengths[i];
 
-            if (totalElementCount > uint.MaxValue)
+            if (totalElementCount > ArrayInfo.MaxArrayLength)
```
@JeremyKuhne this is a limitation I was supposed to introduce a long time ago: just don't allow for multi-dimensional arrays with the number of elements being `> Array.MaxLength`

JeremyKuhne replied:
Should `totalElementCount` be `checked`?

adamsitnik replied:
https://github.com/dotnet/runtime/blob/a3365e4c442a728bc82e59d5583cf83bc6d31d72/src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/BinaryArrayRecord.cs#L135-L139

https://github.com/dotnet/runtime/blob/a3365e4c442a728bc82e59d5583cf83bc6d31d72/src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/ArrayInfo.cs#L46-L53

In worst case scenario, `totalElementCount` will be `Array.MaxLength - 1` and `lengths[i]` will be `Array.MaxLength`.

It does not overflow `Int64`:

```log
> checked((long)(Array.MaxLength - 1) * (long)Array.MaxLength)
4611685771466771690
```

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/NrbfDecoder.cs:
```diff
@@ -69,28 +69,21 @@ public static bool StartsWithPayloadHeader(Stream stream)
-                if (read == 0)
-                    throw new EndOfStreamException();
-                offset += read;
+                return false;
```
the goal of this change is to simply return `false` rather than throw `EOSE`

lonitra replied:
Do we need to set the stream position back to the beginning before we return here?

adamsitnik replied:
> Do we need to set the stream position back to the beginning before we return here?

Great catch!

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/RecordMap.cs:
```diff
@@ -56,7 +56,7 @@ internal void Add(SerializationRecord record)
                 }
 #endif
-                throw new SerializationException(SR.Format(SR.Serialization_DuplicateSerializationRecordId, record.Id));
+                throw new SerializationException(SR.Format(SR.Serialization_DuplicateSerializationRecordId, record.Id._id));
```
`SerializationRecordId` does not override `ToString`, we need to specify the Id in explicit way

JeremyKuhne replied:
Should it override ToString?

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/RectangularArrayRecord.cs:
```diff
@@ -24,18 +24,8 @@ private RectangularArrayRecord(Type elementType, ArrayInfo arrayInfo,
-            : new LinkedList<object>();
-#endif
-
+        // ArrayInfo.GetSZArrayLength ensures to return a value <= Array.MaxLength
```
Since now we don't allow for any MD array with number of elements `> Array.MaxLength`, we don't need to use `LinkedList` anymore

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/RectangularArrayRecord.cs:
```diff
@@ -176,7 +170,11 @@ internal static RectangularArrayRecord Create(BinaryReader reader, ArrayInfo arr
                 PrimitiveType.Double => sizeof(double),
-                _ => -1
+                PrimitiveType.TimeSpan => sizeof(ulong),
+                PrimitiveType.DateTime => sizeof(ulong),
```
this is an improvement that I got into when I was making all the `switch` statements throw in the `default` case, we were not handling `TimeSpan` and `DateTime` properly before

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/RectangularArrayRecord.cs:
```diff
@@ -155,8 +146,11 @@ internal static RectangularArrayRecord Create(BinaryReader reader, ArrayInfo arr
             BinaryType.Object => typeof(object),
-            _ => typeof(ClassRecord)
+            BinaryType.ObjectArray => typeof(object[]),
+            BinaryType.Class or BinaryType.SystemClass => typeof(ClassRecord),
```
same as below. Moreover, if this logic will change with https://github.com/dotnet/runtime/pull/107627, but I would prefer to merge this PR first (it's simpler and has no breaking changes)

adamsitnik commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/SerializationRecordId.cs:
```diff
@@ -15,6 +16,7 @@ namespace System.Formats.Nrbf;
 /// <summary>
 /// The ID of <see cref="SerializationRecord" />.
 /// </summary>
+[DebuggerDisplay("{_id}")]
```
It's just user experience improvement

adamsitnik committed:
revert a change that is going to get a fix in separate PR as it's not complete, but increases the amount of code that needs to be reviewed

adamsitnik committed:
fuzzer: throw SerializationException when PrimitiveType.String is provided in a place where it's not valid

adamsitnik commented:
@MihuBot fuzz NrbfDecoder

adamsitnik committed:
address Loni feedback: restore the Stream position when encountering EOF

adamsitnik commented:
@MihuBot fuzz NrbfDecoder

bartonjs commented on src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/MemberTypeInfo.cs:
```diff
@@ -53,10 +53,14 @@ internal static MemberTypeInfo Decode(BinaryReader reader, int count, PayloadOpt
+                    // These types have no additional data.
                     break;
+                default:
+                    throw new InvalidOperationException();
```
FWIW, InvalidOperationException is for when a method is invoked on an object and the object has state that doesn't support it.  For example, calling Stream.Read when Stream.CanRead says false.  The object is not in a state where you can call the method.

Many of yours are probably more likely InvalidDataException, where the default message is "Found invalid data while decoding."

bartonjs replied:
Immo corrected me.  Stream.Read when Stream.CanRead is false is NotSupportedException (you should have checked the property on the stream).  InvalidOperationException is for controllable state that was controlled wrong, like an object where you have to set 3 properties before calling "DoStuff", and you called DoStuff without setting all 3.

But this is still most likely InvalidDataException, because the operation made sense to do, just the data that it read back was not internally consistent.

adamsitnik replied:
What I wanted to achieve is:
- ensure all methods that parse enums throw SerializationException when they encounter invalid value (out of range, not in the spec)
- assume that the above is not true and always handle all enum values in switch statements in explicit way, throw in the default case (so if there is a bug in parsing logic, we don't get undefined behavior but just an exception) 

My initial idea was to use `UnreachableException`, but it's not part of NS2.0. I decided to use `InvalidOperationException` and I am open to changing it.

It's being tested by the Fuzzer which does not catch `InvalidOperationException` and treats it as a bug:

https://github.com/dotnet/runtime/blob/4930e1b44027484e13a29c7f3230e1f0bb397e7c/src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/NrbfDecoderFuzzer.cs#L100-L106

(as shown in the fuzzer comments in this PR that has discovered places where it was missing)

adamsitnik committed:
Remove "Null" and "String" from the internal enum PrimitiveType as it is always illegal to parse it and every switch needed to take this into account.

Now we reject those values at parse time and don't need to handle them later.

- In case of BinaryArray (NRBF 2.4.3.1):
 "If the BinaryTypeEnum value is Primitive, the PrimitiveTypeEnumeration
 value in AdditionalTypeInfo MUST NOT be Null (17) or String (18)."
- In case of MemberPrimitiveTyped (NRBF 2.5.1):
 "PrimitiveTypeEnum (1 byte): A PrimitiveTypeEnumeration
 value that specifies the Primitive Type of data that is being transmitted.
 This field MUST NOT contain a value of 17 (Null) or 18 (String)."
- In case of ArraySinglePrimitive (NRBF 2.4.3.3):
 "A PrimitiveTypeEnumeration value that identifies the Primitive Type
 of the items of the Array. The value MUST NOT be 17 (Null) or 18 (String)."
- In case of MemberTypeInfo (NRBF 2.3.1.2):
 "When the BinaryTypeEnum value is Primitive, the PrimitiveTypeEnumeration
 value in AdditionalInfo MUST NOT be Null (17) or String (18)."

adamsitnik commented:
@MihuBot fuzz NrbfDecoder
